lint/long-chain_checker.go

```diff
@@ -0,0 +1,139 @@
+package lint
+
+import (
+	"fmt"
+	"go/ast"
+	"strings"
+
+	"github.com/Quasilyte/astcmp"
+)
+
+type longChainChecker struct {
+	ctx *Context
+}
+
+func newLongChainChecker(ctx *Context) Checker {
+	return &longChainChecker{
+		ctx: ctx,
+	}
+}
+
+// Check runs long-chain checker for f.
+//
+// Features
+//
+// Find long expression chains that are repeated more than N times
+// and consist of multiple selector/index expressions.
+func (c *longChainChecker) Check(f *ast.File) {
+	for _, decl := range f.Decls {
+		if decl, ok := decl.(*ast.FuncDecl); ok {
+			if decl.Body == nil {
+				continue
+			}
+			for _, stmt := range decl.Body.List {
+				if stmt, ok := stmt.(*ast.SwitchStmt); ok {
+					c.checkSwitch(stmt)
+				}
+			}
+		}
+	}
+}
+
+// The general idea is to find greatest common prefix of case expression
+// statements. To do that, we split each case expression by "." (only for
+// SelectorExpr) and compare each ith element of every case expression.
+// If they all equal, we continue algorithm on i+1. If they are not
+// we assume prefix length = i-1
+// If prefix length greater then n we create warning
+//
+// TODO: warn if at least m case expressions have common prefix with
+// length > n
+
```

Comment:
This empty line makes comment detached, so it's not a proper doc comment from AST/godoc point of view.
It's OK for this revision. We can fix that later. This review is already gone for too long.

Reply:
It's better to have misplaced comment than no documentation at all. :)